hi-ml-histopathology/testhisto/testhisto/utils/test_output_utils.py

```diff
@@ -0,0 +1,133 @@
+from pathlib import Path
+from typing import Dict, List, Mapping
+from unittest.mock import MagicMock
+
+import pytest
+from ruamel.yaml import YAML
+from torchmetrics.metric import Metric
+
+from histopathology.utils.output_utils import DeepMILOutputsHandler, OutputsPolicy
+from histopathology.utils.naming import MetricsKey
+
+_PRIMARY_METRIC_KEY = MetricsKey.ACC
+
+
+def _create_outputs_policy(outputs_root: Path) -> OutputsPolicy:
+    return OutputsPolicy(outputs_root=outputs_root,
+                         primary_val_metric=_PRIMARY_METRIC_KEY,
+                         maximise=True)
+
+
+def _create_outputs_handler(outputs_root: Path) -> DeepMILOutputsHandler:
+    return DeepMILOutputsHandler(
+        outputs_root=outputs_root,
+        n_classes=1,
+        tile_size=224,
+        level=1,
+        slides_dataset=None,
+        class_names=None,
+        primary_val_metric=_PRIMARY_METRIC_KEY,
+        maximise=True,
+    )
+
+
+def _get_mock_metrics_dict(value: float) -> Dict[MetricsKey, Metric]:
+    mock_metric = MagicMock()
+    mock_metric.compute.return_value = value
+    return {_PRIMARY_METRIC_KEY: mock_metric}
+
+
+def test_outputs_policy_persistence(tmp_path: Path) -> None:
+    initial_epoch = 0
+    initial_value = float('-inf')
+
+    # New handler should match initial settings
+    policy = _create_outputs_policy(tmp_path)
+    assert policy._best_metric_epoch == initial_epoch
+    assert policy._best_metric_value == initial_value
+
+    # Recreating a handler should recover the same (arbitrary) settings
+    arbitrary_epoch = 42
+    arbitrary_value = 0.123
+    policy._best_metric_epoch = arbitrary_epoch
+    policy._best_metric_value = arbitrary_value
+    policy._save_best_metric()
+
+    reloaded_policy = _create_outputs_policy(tmp_path)
+    assert reloaded_policy._best_metric_epoch == arbitrary_epoch
+    assert reloaded_policy._best_metric_value == arbitrary_value
+
+    # Handler re-creation should fail if primary metric name differs from what is saved
+    wrong_metric_name = 'wrong_metric_name'
+    yaml = YAML()
+    contents = yaml.load(policy.best_metric_file_path)
+    contents[OutputsPolicy._PRIMARY_METRIC_KEY] = wrong_metric_name
+    yaml.dump(contents, policy.best_metric_file_path)
+
+    with pytest.raises(ValueError) as e:
+        _create_outputs_policy(tmp_path)
+    assert wrong_metric_name in str(e.value)
+
+    # If the best-metric file is missing, a new handler should have a fresh initialisation
+    policy.best_metric_file_path.unlink()  # delete best metric file
+
+    fresh_policy = _create_outputs_policy(tmp_path)
+    assert fresh_policy._best_metric_epoch == initial_epoch
+    assert fresh_policy._best_metric_value == initial_value
+
+
+def test_overwriting_val_outputs(tmp_path: Path) -> None:
```

Comment:
the outputs handler is pretty complex, and writes tons of files, figures, etc. Would it be feasible to add asserts on those? Checking that files look like they should look, etc

Reply:
The individual plotting functions imported from `metrics_utils` already have tests for file contents in `test_metrics_utils.py`.